### PR TITLE
Feature/coap client

### PR DESCRIPTION
**Pull request by danaivach**
* Update `Form` : 
  - returns subprotocol and method name (if not present) based on default protocol bindings (HTTP/CoAP)
* Update `TDGraphReader`:
  - Reads HTTP/CoAP method names based on the URI scheme of a target 
* Update `TDGraphWriter`:
  - Writes subprotocol as string (e.g. "longpoll") or IRI (e.g. cov:observe)
  - Writes HTTP/CoAP method names based on the URI scheme of a target 
* Create `TDCoAPRequest` for constructing and executing a CoAP request:
  - Method `TDCoapResponse execute()` : synchronous requests
  - Method `void execute(TDCoapHandler handler)` : asynchronous requests, the` handler` is invoked upon notification
  - Method `TDCoapObserveRelation establishRelation(TDCoapHandler handler)` : asynchronous observe requests, the `handler` is invoked upon notification, the observe relation can be monitored/manipulated through a `TDCoapObserveRelation`. Requires `cov:observe`.
  - Method `TDCoapObserveRelation establishRelationAndWait(TDCoapHandler handler)` : synchronous observe requests, waits until the relation is established, the `handler` is invoked upon notification, the observe relation can be monitored/manipulated through a `TDCoapObserveRelation`. Requires `cov:observe`.
* Create `TDCoAPResponse`, it wraps a CoAP response.
* Create abstract` TDCoapHandler ` for reacting to notifications from asynchronous CoAP requests and established CoAP observe relations.
* Create `TDCoapObserveRelation`, it wraps a control-handle of a CoAP observe relation.

### CODE REVIEW

File: src/main/java/ch/unisg/ics/interactions/wot/td/io/TDGraphWriter.java [hunk 1]
```diff
@@ -1,207 +1,205 @@
...
-      if (form.getMethodName().isPresent() && form.getOperationTypes().size() == 1) {
-        graphBuilder.add(formId, rdf.createIRI(HTV.methodName), form.getMethodName().get());
+
+      // Only writes the method name for forms with one operation type (to avoid ambiguity)
+      if (form.getOperationTypes().size() == 1) {
+        if (Arrays.stream(HTTP_URI_SCHEMES).anyMatch(form.getTarget()::contains)) {
+          graphBuilder.add(formId, rdf.createIRI(HTV.methodName), form.getMethodName().get());
+        } else if (Arrays.stream(COAP_URI_SCHEMES).anyMatch(form.getTarget()::contains)) {
+          graphBuilder.add(formId, rdf.createIRI(COV.methodName), form.getMethodName().get());
+        }
```
Comment (danaivach) [inline src/main/java/ch/unisg/ics/interactions/wot/td/io/TDGraphWriter.java:198]: Write HTTP/CoAP method names based on the URI scheme of a target

File: src/main/java/ch/unisg/ics/interactions/wot/td/io/TDGraphWriter.java [hunk 2]
```diff
@@ -210,14 +208,19 @@ private void addFormsForInteraction(Resource interactionId, InteractionAffordanc
...
-        graphBuilder.add(formId, rdf.createIRI(HCTL.forSubProtocol), subprotocol.get());
+
+      Optional<String> subProt = form.getSubprotocol();
+      if (subProt.isPresent()) {
+        try {
+          IRI subProtIri = rdf.createIRI(subProt.get());
+          graphBuilder.add(formId, rdf.createIRI(HCTL.forSubProtocol), subProtIri);
+        } catch (IllegalArgumentException e) {
+          graphBuilder.add(formId, rdf.createIRI(HCTL.forSubProtocol), subProt.get());
+        }
```
Comment (danaivach) [inline src/main/java/ch/unisg/ics/interactions/wot/td/io/TDGraphWriter.java:219]: Writes subprotocol as string (e.g. "longpoll") or IRI (e.g. cov:observe)

File: src/main/java/ch/unisg/ics/interactions/wot/td/io/TDGraphReader.java [hunk 3]
```diff
@@ -265,95 +251,104 @@ private ActionAffordance readAction(Resource affordanceId) {
...
-      
-      Optional<Literal> subprotocolOpt = Models.objectLiteral(model.filter(formId, 
-          rdf.createIRI(HCTL.forSubProtocol), null));
-      
-      Set<IRI> opsIRIs = Models.objectIRIs(model.filter(formId, rdf.createIRI(HCTL.hasOperationType), 
-          null));
+      } else if (Arrays.stream(COAP_URI_SCHEMES).anyMatch(targetOpt.toString()::contains)) {
+        methodNameOpt = Models.objectLiteral(model.filter(formId,
+          rdf.createIRI(COV.methodName), null));
+      } else {
```
Comment (danaivach) [inline src/main/java/ch/unisg/ics/interactions/wot/td/io/TDGraphReader.java:313]: Read HTTP/CoAP method names based on the URI scheme of a target

**Comment by danaivach** [comment]
Even if you cancel a `TDCoapObserveRelation`, the program never terminates, unless you call `shutdownExecutors()` on your `TDCoapRequest` instance. This is because ["the (CoAP) client executor uses a user thread (not a daemon thread) that needs to be stopped to exit the program by calling {@link #shutdown()}"](https://github.com/eclipse/californium/blob/master/californium-core/src/main/java/org/eclipse/californium/core/CoapClient.java).

**Review by andreiciortea — APPROVED** [review]
Thanks @danaivach, great!

We should refactor `ProtocolBinding` to move all default mappings to protocol-specific locations (rather than maintaining one class with mappings for all). We can also refactor the `clients` package into 2 sub-packages for HTTP and CoAP.

I will merge this PR and will open an issue for the refactoring tasks.

**Pull request merged**